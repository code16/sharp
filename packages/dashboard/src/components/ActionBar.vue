<template>
    <ActionBar>
        <template v-slot:extras>
            <div class="row mx-n2">
                <template v-for="filter in filters">
                    <div class="col-auto px-2">
                        <FilterDropdown
                            :filter="filter"
                            :value="filterValue(filter.key)"
                            @input="handleFilterChanged(filter, $event)"
                            :key="filter.id"
                        />
                    </div>
                </template>
            </div>
        </template>
        <template v-if="commands.length" v-slot:extras-right>
            <CommandsDropdown class="SharpActionBar__actions-dropdown SharpActionBar__actions-dropdown--commands"
                :commands="commands"
                @select="handleCommandSelected"
            >
                <template v-slot:text>
                    {{ l('dashboard.commands.dashboard.label') }}
                </template>
            </CommandsDropdown>
        </template>
    </ActionBar>
</template>

<script>
    import { mapGetters } from 'vuex';
    import { ActionBar } from 'sharp-ui';
    import { FilterDropdown } from 'sharp-filters';
    import { CommandsDropdown } from 'sharp-commands';
    import { Localization } from "sharp/mixins";

    export default {
        name: 'SharpActionBarDashboard',
        mixins: [Localization],
        components: {
            ActionBar,
            FilterDropdown,
            CommandsDropdown,
        },
        props: {
            commands: Array,
        },
        computed: {
            ...mapGetters('dashboard', {
                filters: 'filters/filters',
                filterValue: 'filters/value',
                filterNextQuery: 'filters/nextQuery',
            })
        },
        methods: {
            filterKey(filter) {
                return `actionbardashboard_${filter.key}`;
            },
            handleFilterChanged(filter, value) {
                this.$router.push({
                    query: {
                        ...this.$route.query,
                        ...this.filterNextQuery({ filter, value }),
                    }
                });
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            }
        }
    }
</script>