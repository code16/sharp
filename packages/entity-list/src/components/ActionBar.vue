<template>
    <ActionBar
        class="SharpActionBarList"
        :class="{'SharpActionBarList--search-active':searchActive}"
        right-class="d-block"
    >
        <template v-slot:left>
            <span class="text-content text-nowrap">{{ count }} {{ l('action_bar.list.items_count') }}</span>
        </template>
        <template v-slot:right>
            <div class="row justify-content-end">
                <template v-if="canSearch && !reorderActive">
                    <div class="col col-lg-auto">
                        <Search
                            class="h-100"
                            :value="search"
                            :active.sync="searchActive"
                            :placeholder="l('action_bar.list.search.placeholder')"
                            @input="handleSearchInput"
                            @submit="handleSearchSubmitted"
                        />
                    </div>
                </template>

                <template v-if="canReorder">
                    <div class="col-auto">
                        <template v-if="reorderActive">
                            <button class="SharpButton SharpButton--secondary-accent h-100" @click="handleReorderButtonClicked">
                                {{ l('action_bar.list.reorder_button.cancel') }}
                            </button>
                            <button class="SharpButton SharpButton--accent h-100" @click="handleReorderSubmitButtonClicked">
                                {{ l('action_bar.list.reorder_button.finish') }}
                            </button>
                        </template>
                        <template v-else>
                            <button class="SharpButton SharpButton--secondary-accent h-100" @click="handleReorderButtonClicked">
                                {{ l('action_bar.list.reorder_button') }}
                            </button>
                        </template>
                    </div>
                </template>

                <template v-if="canCreate && !reorderActive">
                    <div class="col-auto">
                        <template v-if="hasForms">
                            <Dropdown class="SharpActionBar__forms-dropdown h-100" :text="l('action_bar.list.forms_dropdown')">
                                <DropdownItem v-for="(form,key) in forms" @click="handleCreateFormSelected(form)" :key="key" >
                                    <ItemVisual :item="form" icon-class="fa-fw"/>{{ form.label }}
                                </DropdownItem>
                            </Dropdown>
                        </template>
                        <template v-else>
                            <button class="SharpButton SharpButton--accent h-100" @click="handleCreateButtonClicked">
                                {{ l('action_bar.list.create_button') }}
                            </button>
                        </template>
                    </div>
                </template>
            </div>
        </template>
        <template v-if="!reorderActive" v-slot:extras>
            <div class="row mx-n2">
                <template v-for="filter in filters">
                    <div class="col-auto px-2">
                        <FilterDropdown
                            :filter="filter"
                            :value="filtersValues[filter.key]"
                            @input="handleFilterChanged(filter, $event)"
                            :key="filter.id"
                        />
                    </div>
                </template>
            </div>
        </template>
        <template v-if="!reorderActive" v-slot:extras-right>
            <template v-if="hasCommands">
                <CommandsDropdown class="SharpActionBar__actions-dropdown SharpActionBar__actions-dropdown--commands"
                    :commands="commands"
                    @select="handleCommandSelected"
                >
                    <template v-slot:text>
                        {{ l('entity_list.commands.entity.label') }}
                    </template>
                </CommandsDropdown>
            </template>
        </template>
    </ActionBar>
</template>

<script>
    import { ActionBar, Dropdown,  DropdownItem, ItemVisual, Search } from 'sharp-ui';
    import { FilterDropdown } from 'sharp-filters';
    import { CommandsDropdown } from 'sharp-commands';

    import { Localization } from 'sharp/mixins';

    export default {
        name: 'SharpActionBarList',

        mixins: [Localization],

        components : {
            ActionBar,
            Dropdown,
            DropdownItem,
            ItemVisual,
            CommandsDropdown,
            FilterDropdown,
            Search,
        },

        props: {
            count: Number,
            search: String,
            filters: Array,
            filtersValues: Object,
            commands: Array,
            forms: Array,

            canCreate: Boolean,
            canReorder: Boolean,
            canSearch: Boolean,

            reorderActive: Boolean
        },

        data() {
            return {
                searchActive: false
            }
        },
        computed: {
            hasForms() {
                return this.forms && this.forms.length > 0;
            },
            hasCommands() {
                return this.commands && this.commands.some(group => group && group.length > 0);
            },
        },
        methods: {
            handleSearchInput(search) {
                this.$emit('search-change', search);
            },
            handleSearchSubmitted() {
                this.$emit('search-submit');
            },
            handleFilterChanged(filter, value) {
                this.$emit('filter-change', filter, value);
            },
            handleReorderButtonClicked() {
                this.$emit('reorder-click');
                document.activeElement.blur();
            },
            handleReorderSubmitButtonClicked() {
                this.$emit('reorder-submit');
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleCreateButtonClicked() {
                this.$emit('create');
            },
            handleCreateFormSelected(form) {
                this.$emit('create', form);
            }
        }
    }
</script>