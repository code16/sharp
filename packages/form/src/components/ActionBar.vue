<template>
    <ActionBar>
        <template v-slot:left>
            <Button type="secondary-accent" @click="handleCancelClicked">
                <template v-if="showBackButton">
                    {{ l('action_bar.form.back_button') }}
                </template>
                <template v-else>
                    {{ l('action_bar.form.cancel_button') }}
                </template>
            </Button>

            <template v-if="showDeleteButton">
                <div class="w-100 h-100">
                    <Collapse transition-class="SharpButton__collapse-transition">
                        <template v-slot:frame-0="{ next }">
                            <button class="SharpButton SharpButton--danger" @click="next(focusDelete)">
                                <svg  width='16' height='16' viewBox='0 0 16 24' fill-rule='evenodd'>
                                    <path d='M4 0h8v2H4zM0 3v4h1v17h14V7h1V3H0zm13 18H3V8h10v13z'></path>
                                    <path d='M5 10h2v9H5zm4 0h2v9H9z'></path>
                                </svg>
                            </button>
                        </template>
                        <template v-slot:frame-1="{ next }">
                            <Button type="danger" @click="handleDeleteClicked" @blur="next()" ref="openDelete">
                                {{ l('action_bar.form.delete_button') }}
                            </Button>
                        </template>
                    </Collapse>
                </div>
            </template>
        </template>
        <template v-slot:right>
            <template v-if="showSubmitButton">
                <Button type="accent" :disabled="uploading" @click="handleSubmitClicked">
                    {{ submitLabel }}
                </Button>
            </template>
        </template>
    </ActionBar>
</template>

<script>
    import { lang } from 'sharp';
    import { ActionBar, Collapse, Button } from 'sharp-ui';
    import { Localization } from "sharp/mixins";

    export default {
        name: 'SharpActionBarForm',
        mixins: [Localization],
        components: {
            ActionBar,
            Collapse,
            Button,
        },
        props: {
            showSubmitButton: Boolean,
            showDeleteButton: Boolean,
            showBackButton: Boolean,
            create: Boolean,
            uploading: Boolean,
        },
        computed: {
            submitLabel() {
                if(this.uploading) {
                    return lang('action_bar.form.submit_button.pending.upload')
                }
                return this.create
                    ? lang('action_bar.form.submit_button.create')
                    : lang('action_bar.form.submit_button.update');
            },
        },
        methods: {
            focusDelete() {
                if(this.$refs.openDelete) {
                    this.$refs.openDelete.focus();
                }
            },
            handleSubmitClicked() {
                this.$emit('submit');
            },
            handleDeleteClicked() {
                this.$emit('delete');
            },
            handleCancelClicked() {
                this.$emit('cancel');
            },
        },
    }
</script>