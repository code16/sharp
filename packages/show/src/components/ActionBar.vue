<template>
    <ActionBar>
        <template v-slot:left>
            <template v-if="showBackButton">
                <a :href="backUrl" class="SharpButton SharpButton--secondary-accent">
                    {{ l('action_bar.show.back_button') }}
                </a>
            </template>
        </template>
        <template v-slot:right>
            <template v-if="canEdit">
                <a :href="formUrl" class="SharpButton SharpButton--accent">
                    {{ l('action_bar.show.edit_button') }}
                </a>
            </template>
        </template>
        <template v-slot:extras-right>
            <div class="row mx-n1">
                <template v-if="hasState">
                    <div class="col-auto px-1">
                        <Dropdown class="SharpActionBar__actions-dropdown SharpActionBar__actions-dropdown--state" :disabled="!canChangeState">
                            <template v-slot:text>
                                <StateIcon :color="state.color" />
                                <span class="text-truncate">{{ state.label }}</span>
                            </template>
                            <template v-for="stateOptions in stateValues">
                                <DropdownItem
                                    @click="handleStateChanged(stateOptions.value)"
                                    :key="stateOptions.value"
                                >
                                    <StateIcon :color="stateOptions.color" />&nbsp;
                                    {{ stateOptions.label }}
                                </DropdownItem>
                            </template>
                        </Dropdown>
                    </div>
                </template>
                <template v-if="hasCommands">
                    <div class="col-auto px-1">
                        <CommandsDropdown class="SharpActionBar__actions-dropdown SharpActionBar__actions-dropdown--commands"
                            :commands="commands"
                            @select="handleCommandSelected"
                        >
                            <template v-slot:text>
                                {{ l('entity_list.commands.instance.label') }}
                            </template>
                        </CommandsDropdown>
                    </div>
                </template>
            </div>
        </template>
    </ActionBar>
</template>

<script>
    import {
        ActionBar,
        Dropdown,
        DropdownItem,
        StateIcon,
    } from 'sharp-ui';

    import {
        CommandsDropdown
    } from 'sharp-commands';

    import { Localization } from "sharp/mixins";

    export default {
        mixins: [Localization],
        components: {
            ActionBar,
            CommandsDropdown,
            Dropdown,
            DropdownItem,
            StateIcon,
        },
        props: {
            commands: Array,
            formUrl: String,
            backUrl: String,
            canEdit: Boolean,
            canChangeState: Boolean,
            showBackButton: Boolean,
            state: Object,
            stateValues: Array,
        },
        computed: {
            hasCommands() {
                return this.commands && this.commands.some(group => group && group.length > 0);
            },
            hasState() {
                return !!this.state;
            },
        },
        methods: {
            handleEditButtonClicked() {
                this.$emit('edit');
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleStateChanged(state) {
                this.$emit('state-change', state);
            },
        }
    }
</script>