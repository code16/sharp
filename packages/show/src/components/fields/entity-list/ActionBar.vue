<template>
    <div class="action-bar">
        <template v-if="hasLeftControls && $slots.default">
            <div class="mb-2">
                <slot />
            </div>
        </template>
        <div class="row action-bar__row">
            <template v-if="hasLeftControls">
                <div class="col-sm action-bar__col">
                    <template v-if="!reorderActive">
                        <div class="row action-bar__row">
                            <template v-for="filter in filters">
                                <div class="col-auto action-bar__col mb-2">
                                    <div class="action-bar__element">
                                        <FilterDropdown
                                            class="h-100"
                                            :filter="filter"
                                            :value="filtersValues[filter.key]"
                                            @input="handleFilterChanged(filter, $event)"
                                            :key="filter.id"
                                        />
                                    </div>
                                </div>
                            </template>
                            <template v-if="canSearch">
                                <div class="col-auto action-bar__col mb-2">
                                    <div class="action-bar__element">
                                        <Search
                                            class="h-100"
                                            :value="search"
                                            :active.sync="searchActive"
                                            :placeholder="l('action_bar.list.search.placeholder')"
                                            @input="handleSearchInput"
                                            @submit="handleSearchSubmitted"
                                        />
                                    </div>
                                </div>
                            </template>
                        </div>
                    </template>
                </div>
            </template>
            <template v-else>
                <div class="col-sm align-self-center action-bar__col mb-2">
                    <slot />
                </div>
            </template>

            <div class="col-sm-auto action-bar__col">
                <div class="row flex-nowrap justify-content-end action-bar__row">
                    <template v-if="canReorder">
                        <div class="col-auto action-bar__col mb-2">
                            <template v-if="reorderActive">
                                <div class="row action-bar__row">
                                    <div class="col-auto action-bar__col">
                                        <button class="SharpButton SharpButton--secondary-accent" @click="handleReorderButtonClicked">
                                            {{ l('action_bar.list.reorder_button.cancel') }}
                                        </button>
                                    </div>
                                    <div class="col-auto action-bar__col">
                                        <button class="SharpButton SharpButton--accent" @click="handleReorderSubmitButtonClicked">
                                            {{ l('action_bar.list.reorder_button.finish') }}
                                        </button>
                                    </div>
                                </div>
                            </template>
                            <template v-else>
                                <button class="SharpButton SharpButton--secondary-accent" @click="handleReorderButtonClicked">
                                    {{ l('action_bar.list.reorder_button') }}
                                </button>
                            </template>
                        </div>
                    </template>
                    <template v-if="canCreate && !reorderActive">
                        <div class="col-auto action-bar__col mb-2">
                            <div class="action-bar__element">
                                <template v-if="hasForms">
                                    <Dropdown class="SharpActionBar__forms-dropdown h-100" :text="l('action_bar.list.forms_dropdown')">
                                        <DropdownItem v-for="(form,key) in forms" @click="handleCreateFormSelected(form)" :key="key" >
                                            <ItemVisual :item="form" icon-class="fa-fw"/>{{ form.label }}
                                        </DropdownItem>
                                    </Dropdown>
                                </template>
                                <template v-else>
                                    <button class="SharpButton SharpButton--accent" @click="handleCreateButtonClicked">
                                        {{ l('action_bar.list.create_button') }}
                                    </button>
                                </template>
                            </div>
                        </div>
                    </template>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
    import { Localization } from 'sharp/mixins';
    import { Search, ItemVisual, Dropdown, DropdownItem } from 'sharp-ui';
    import { FilterDropdown } from 'sharp-filters';

    export default {
        mixins: [Localization],
        components: {
            ItemVisual,
            Search,
            Dropdown,
            DropdownItem,
            FilterDropdown,
        },
        props: {
            count: Number,
            search: String,
            filters: Array,
            filtersValues: Object,
            commands: Array,
            forms: Array,

            canCreate: Boolean,
            canReorder: Boolean,
            canSearch: Boolean,

            reorderActive: Boolean
        },
        data() {
            return {
                searchActive: false,
            }
        },
        computed: {
            hasForms() {
                return this.forms && this.forms.length > 0;
            },
            hasLeftControls() {
                const filters = this.filters || [];
                return this.count > 0 && (filters.length > 0 || this.canSearch);
            },
        },
        methods: {
            handleSearchInput(search) {
                this.$emit('search-change', search);
            },
            handleSearchSubmitted() {
                this.$emit('search-submit');
            },
            handleFilterChanged(filter, value) {
                this.$emit('filter-change', filter, value);
            },
            handleReorderButtonClicked() {
                this.$emit('reorder-click');
                document.activeElement.blur();
            },
            handleReorderSubmitButtonClicked() {
                this.$emit('reorder-submit');
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleCreateButtonClicked() {
                this.$emit('create');
            },
            handleCreateFormSelected(form) {
                this.$emit('create', form);
            },
        }
    }
</script>