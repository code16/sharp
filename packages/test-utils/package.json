{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}