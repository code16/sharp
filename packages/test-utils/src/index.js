export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}