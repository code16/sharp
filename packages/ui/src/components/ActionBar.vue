<template>
    <div class="SharpActionBar" v-if="ready">
        <div class="SharpActionBar__bar">
            <div class="container">
                <div class="row mx-n2">
                    <div class="col left px-2 my-1 my-sm-0">
                        <slot name="left"></slot>
                    </div>
                    <div class="col right px-2 my-1 my-sm-0" :class="rightClass">
                        <slot name="right"></slot>
                    </div>
                </div>
            </div>
        </div>
        <div :class="{ 'container':container }">
            <div class="row">
                <div class="col">
                    <div class="SharpActionBar__extras">
                        <slot name="extras" />
                    </div>
                </div>
                <template v-if="$slots['extras-right']">
                    <div class="col-auto">
                        <div class="SharpActionBar__extras">
                            <slot name="extras-right" />
                        </div>
                    </div>
                </template>
            </div>
        </div>
    </div>
</template>

<script>
    export default {
        name: 'SharpActionBar',
        props: {
            ready: {
                type: Boolean,
                default: true,
            },
            container: Boolean,
            rightClass: String,
        }
    }
</script>