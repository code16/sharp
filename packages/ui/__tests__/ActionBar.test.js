import { mount } from '@vue/test-utils';
import ActionBar from '../src/components/ActionBar.vue';

describe('action-bar', ()=>{
    test('can mount non ready ActionBar', ()=>{
        expect(mount(ActionBar, { propsData: { ready: false } }).html()).toMatchSnapshot();
    });
    test('can mount ready ActionBar', ()=>{
        expect(mount(ActionBar, { propsData: { ready: true } }).html()).toMatchSnapshot();
    });
    test('can mount with slots ActionBar', ()=>{
        expect(mount(ActionBar, {
            propsData: { ready: true },
            slots: {
                left: '<div>LEFT CONTENT</div>',
                right: '<div>RIGHT CONTENT</div>',
                extras: '<div>EXTRAS CONTENT</div>'
            }
        }).html()).toMatchSnapshot();
    });
});