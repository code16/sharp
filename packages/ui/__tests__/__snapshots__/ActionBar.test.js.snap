// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action-bar can mount non ready ActionBar 1`] = `undefined`;

exports[`action-bar can mount ready ActionBar 1`] = `
<div class="SharpActionBar">
	<div class="SharpActionBar__bar">
		<div class="container">
			<div class="row mx-n2">
				<div class="col left px-2 my-1 my-sm-0"></div>
				<div class="col right px-2 my-1 my-sm-0"></div>
			</div>
		</div>
	</div>
	<div class="">
		<div class="row">
			<div class="col">
				<div class="SharpActionBar__extras"></div>
			</div>
			<!---->
		</div>
	</div>
</div>
`;

exports[`action-bar can mount with slots ActionBar 1`] = `
<div class="SharpActionBar">
	<div class="SharpActionBar__bar">
		<div class="container">
			<div class="row mx-n2">
				<div class="col left px-2 my-1 my-sm-0">
					<div>LEFT CONTENT</div>
				</div>
				<div class="col right px-2 my-1 my-sm-0">
					<div>RIGHT CONTENT</div>
				</div>
			</div>
		</div>
	</div>
	<div class="">
		<div class="row">
			<div class="col">
				<div class="SharpActionBar__extras">
					<div>EXTRAS CONTENT</div>
				</div>
			</div>
			<!---->
		</div>
	</div>
</div>
`;
